packages/client/src/hooks/useForm.tsx

```diff
@@ -0,0 +1,86 @@
+import React, { useCallback, useState } from 'react'
+import { Field } from '../components/UI/Field'
+import { FieldType } from '../types/field'
+
+export const useForm = (config: FieldType[]): any[] => {
+  const initialData: Record<string, any> = {}
+  const initialValues: Record<string, any> = {}
+  for (const field of config) {
+    initialData[field.name] = field
+    initialValues[field.name] = field.value
+  }
+  const [formData, setFormData] = useState(initialData)
+
+  const handleChange = (e: React.FormEvent<HTMLInputElement>) => {
+    const name = e.currentTarget.name
+    const value = e.currentTarget.value
+    const fieldData = { ...formData[name] }
+
+    if (!fieldData.isValid) {
+      fieldData.isValid = fieldData.rule.test(fieldData.value)
+    }
+
+    fieldData.value = value
+
+    setFormData({
+      ...formData,
+      [name]: fieldData
+    })
+  }
+
+  const formValues = useCallback(() => {
+    const values: Record<string, any> = {}
+    const fields = Object.keys(formData)
+
+    for (const i of fields) {
+      values[i] = formData[i].value
+    }

```

Comment:
```suggestion
    for (const field of fields) {
      values[field] = formData[field].value
    }
```

В данном случае i, это не индекс, а [элемент](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of). Поэтому лучше назвать его не i так как складывается ложное ощущение, что это индекс)) 
Нужно называть переменки по смыслу, везде.

Reply:
тут переписала, поняла что зря Object.keys делала